### ISSUE TITLE
Inconsistent behavior: MyPy rejects isinstance with Any, but accepts cast from Any for ignored imports

### ISSUE THREAD

**Issue by carlos-superior** [python/mypy#17844]
**Context**

When working with the `cassandra.cluster` module in a Python project, i had to use `type: ignore` on the import because there are no compatible MyPy stubs available for the module. 

Below is a simplified version of how the code is set up.

`from cassandra.cluster import Session as CassandraConnection` # type: ignore (mypy will treat it just like Any) 
`from typing import TypeAlias, cast, Any`

`TransactionalConnection: TypeAlias = CassandraConnection`
 
This approach generally works, but it causes issues with type checking in MyPy.

---

**Initial Problem**

When trying to use `isinstance()` to check if a variable is an instance of `TransactionalConnection`, Mypy raises the following error: 

_`Cannot use isinstance() with Any type  [misc]`_

This happens because MyPy is treating `CassandraConnection` as `Any` due to the `#type: ignore`. Here's the code that triggers the error: 

`if isinstance(db_client, TransactionalConnection):
 pass
 ` -> error occurs here: Cannot use isinstance() with Any type.

---

**Paradox**

Interestingly, when using `cast(Any, CassandraConnection)`, Mypy accepts the typing without issues. In other words, when i cast `Any` to `CassandraConnection`, Mypy does not raise any errors, even though the underlying type is still `Any`. Here's the example: 

`TransactionalConnection: TypeAlias = cast(Any, CassandraConnection)` -> This works with no MyPy errors.

This behavior is paradoxical because, in both cases: (`isinstance` and `cast`), the underlying type is `Any`, but Mypy handles them differently.

---

**Full Scenario Example**

```
# Core
import json
import os
from typing import TypedDict, TypeAlias, cast, Any

# Libraries
from psycopg2.extensions import connection as PostgreSQLConnection
import psycopg2
from psycopg2 import sql as psycopg2_sql
from cassandra.cluster import Session as CassandraConnection    # type: ignore
from cassandra.auth import PlainTextAuthProvider    # type: ignore
from cassandra.cluster import Cluster

# Core functionality
sql = psycopg2_sql
SqlCursor = psycopg2.extensions.cursor

# Defining type aliases
RelationalConnection: TypeAlias = PostgreSQLConnection
TransactionalConnection: TypeAlias = cast(Any, CassandraConnection)

class DbConnections(TypedDict):
    transactional: TransactionalConnection
    relational: RelationalConnection
```

Here’s the part where the `isinstance` check causes the error:

```
if isinstance(db_client, TransactionalConnection):  
    pass
```
error: Cannot use isinstance() with Any type

**But when using** `cast`, **MyPy doesn't raise any errors**:
`TransactionalConnection: TypeAlias = cast(Any, CassandraConnection)`

--- 

**Conclusion**

The inconsistent behavior between using `isinstance()` and `cast` when dealing with the `Any` type seems to be a bug or, at the very least, an inconsistency in Mypy. If `cast()` accepts `Any` without issue, `isinstance()` should behave similarly, as both deal with the `Any` type. However, Mypy rejects `isinstance`, but accepts `cast`. 

This inconsistent behavior can confuse developers working with modules that lack proper typing, requiring the use of `# type: ignore`. I hope this issue can be reviewed, and Mypy can offer consistent behavior between `isinstance` and cast. 

--- 

**Environment**

- Python version 3.12.3 
- Flags: `--strict`
- mypy version: 1.11.0
- mypy-extensions version: 1.0.0

**Comment by brianschubert** [comment]
> In other words, when i cast `Any` to `CassandraConnection`, Mypy does not raise any errors, even though the underlying type is still `Any`. Here's the example: [...] `cast(Any, CassandraConnection)`

I haven't looked at this issue closely, but just in case it was a source of confusion, I wanted to point out that `cast(Any, CassandraConnection)` is a cast _to_ `Any`, not _from_ `Any`. The signature for cast is `cast(<assumed type>, <runtime value>)`. When you write `cast(Any, CassandraConnection)`, you get back the runtime value `CassandraConnection`, but ask static type checkers to treat it as though it were an `Any`.

> If `cast()` accepts Any without issue, `isinstance()` should behave similarly

There's a big difference between `cast` and `isinstance`, which is that `cast` only affects static type checking whereas `isinstance` also has runtime behavior. Mypy rejects using `Any` with `isinstance` because combing the two at runtime produces a `TypeError`:
```pycon
>>> isinstance("foo", Any)
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    isinstance("foo", Any)
    ~~~~~~~~~~^^^^^^^^^^^^
  File "/usr/lib/python3.13/typing.py", line 589, in __instancecheck__
    raise TypeError("typing.Any cannot be used with isinstance()")
TypeError: typing.Any cannot be used with isinstance()
```

**Comment by wyattscarpenter** [comment]
I also don't find the behavior paradoxical, for the reasons explained above.

Interestingly, using CassandraConnection directly will not have this problem. And if you just assign it to something instead of using TypeAlias, you also will not have this problem. Even though it's still Any.

Gist URL: https://gist.github.com/mypy-play/9523cccdc8cfb1a8dcbbda64c941073f
Playground URL: https://mypy-play.net/?mypy=master&python=3.12&flags=strict&gist=9523cccdc8cfb1a8dcbbda64c941073f
